contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c

```diff
@@ -1848,12 +1848,14 @@ PrepareRowDescription(TupleDesc typeinfo, PlannedStmt *plannedstmt, List *target
 			case TDS_SEND_VARCHAR:
 				SetColMetadataForCharTypeHelper(col, TDS_TYPE_VARCHAR,
 												att->attcollation, (atttypmod == -1) ?
-												atttypmod : (atttypmod - 4));
+												(tdsVersion >= TDS_VERSION_7_2 ? atttypmod : 0x7fff) :
+												(atttypmod - 4));
 				break;
 			case TDS_SEND_NVARCHAR:
 				SetColMetadataForCharTypeHelper(col, TDS_TYPE_NVARCHAR,
 												att->attcollation, (atttypmod == -1) ?
-												atttypmod : (atttypmod - 4) * 2);
+												(tdsVersion >= TDS_VERSION_7_2 ? atttypmod : 0x7ffe) :
```

Comment:
> 1. what is meaning of 0x7ffe and 0x7fff?

Apologies, these constant values were incorrect (they are ignored by JTDS), correct value for them is `0xfffe` (max size for `BIGVARCHARTYPE` and `NVARCHARTYPE` from [2.2.5.4.3 Variable-Length Data Types](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/ce3183a6-9d89-47e8-a02f-de5a1a1303de)). I've updated the constants to this value.


>     2. IIUC, we are still able to create table with [n]varchar(max) but we can only send 0x7fff amount of data (truncated)? Have we considered this scenario where you tried to insert the data of length > 0x7fff and try to retrieve it through JTDS driver?

Currently Babelfish will return an error on the attempt to insert or select `varchar(max)` value with length exceeding `0xfffe`. AFAIU the root cause here is the lack of actual support for TDS 7.1 types handling in Babelfish (that is still advertised in [its docs](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/babelfish.html)). Babelfish uses the following TDS types for corresponding T-SQL types:

 - `varchar(max)` - `BIGVARCHARTYPE`
 - `nvarchar(max)` - `NVARCHARTYPE`
 - `varbinary(max)` - `BIGVARBINARYTYPE`

And for these types Babelfish uses PLP (Partially Length-prefixed Bytes, [2.2.5.2.3 Data Type Dependent Data Streams](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/3f983fde-0509-485a-8c40-a9fa6679a828)) that was introduced in TDS 7.2 (and thus not supported in JTDS) and that allows to handle size larger than `0xfffe` for these types.

When MSSQL is run with JTDS using TDS 7.1 protocol it uses the following types mapping instead (without PLP):
 
 - `varchar(max)` - `TEXTTYPE`
 - `nvarchar(max)` - `NTEXTTYPE`
 - `varbinary(max)` - `IMAGETYPE`

And these types allow to handle size up to `0x7fffffff` without PLP.

I am not sure how to proceed with this problem - whether any changes to `babelfishpg_tds` regarding this type mapping would be acceptable or whether it is acceptable to have a limited JTDS support with max length of `varchar/nvarchar/varbinary(max)` limited to `65534`. This point it not addressed in the updated patch.

> 
>     3. don't we need to handle varbinary in the same way?

Yes, added the same max length handling for `varbinary(max)` in updated patch.

Reply:
For the point no 2, 

1. **Currently Babelfish will return an error on the attempt to insert or select varchar(max) value with length exceeding 0xfffe. AFAIU the root cause here is the lack of actual support for TDS 7.1 types handling in Babelfish** -- From where are getting this error? can you please some more information so that we can investigate from our side also.
2. **About handling datatypes specially for different TDS version** -- It is possible to handle [n]varchar(max) and varbinary(max) in a special way based on TDS protocol. For the reference, you can check how have we handled data types such as DATE for the TDS version < 7.3A. We conditionally set metadata[1] and call the sender function of nvarchar[2] to return the output. so we can follow the same here in this case also. 

But before doing that, lets come up with all the gaps first and address them in this pr only. 

Just small suggestion: Please try to use github link for the Microsoft documentation. 

[1]: https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_4_X_DEV/contrib/babelfishpg_tds/src/backend/tds/tdsresponse.c#L1128
[2]: https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_4_X_DEV/contrib/babelfishpg_tds/src/backend/tds/tdstypeio.c#L3134

Reply:
seems like mapping is not yet addressed in this PR? 

varchar(max) - TEXTTYPE
nvarchar(max) - NTEXTTYPE
varbinary(max) - IMAGETYPE

Reply:
No, these TDS types usage is not implemented yet. The intention is to add a working test harness first, got the tests for the failures (some of them should be fixes by thee new types, but there are others in the same area - described in comments below) and then proceed with code changes. If the approach with JDBC  tests with `useJTDSInsteadOfMSSQLJDBC` looks like a right direction - I will proceed with actual tests and code changes one by one.

Reply:
Implemented the usage of  text/ntext/image column types for varchar(max)/nvarchar(max)/varbinary(max) results. Will write more details on remaining failures in a comment below.